extensions/functions/src/main/java/io/crate/window/OffsetValueFunctions.java

```diff
@@ -45,34 +46,173 @@
+        }
+
+        public Integer findNonNullOffsetFromCurrentIndex(int idxInPartition,
+                                                         int offset,
+                                                         WindowFrameState currentFrame,
+                                                         List<? extends CollectExpression<Row, ?>> expressions,
+                                                         Input[] args) {
+            /* Search for 'offset' number of non-null elements in 'offsetDirection' and return the index if found.
+               If index goes out of bound, an exception will be thrown and eventually invoke getDefaultOrNull(...) */
+            for (int i = 1, counter = 0; ; i++) {
+                Object value = getValueAtTargetIndex(getTargetIndex(idxInPartition, i),
+                                                     currentFrame,
+                                                     expressions,
+                                                     args);
+                if (value != null) {
+                    counter++;
+                    if (counter == offset) {
+                        return getTargetIndex(idxInPartition, i);
+                    }
+                }
+            }
+        }
+
+        public Object getValueAtOffset(int idxAtPartition,
+                                       int offset,
+                                       WindowFrameState currentFrame,
+                                       List<? extends CollectExpression<Row, ?>> expressions,
+                                       boolean ignoreNulls,
+                                       Input[] args) {
+            if (ignoreNulls == false) {
+                var targetIndex = getTargetIndex(idxAtPartition, offset);
+                return getValueAtTargetIndex(targetIndex, currentFrame, expressions, args);
+            } else {
+                return getValueAtOffsetIgnoringNulls(idxAtPartition, offset, currentFrame, expressions, args);
+            }
+        }
+
+        static Object getValueAtTargetIndex(Integer targetIndex,
+                                            WindowFrameState currentFrame,
+                                            List<? extends CollectExpression<Row, ?>> expressions,
+                                            Input[] args) {
+            if (targetIndex == null) {
+                throw new IndexOutOfBoundsException();
+            }
+            Object[] rowCells = currentFrame.getRowInPartitionAtIndexOrNull(targetIndex);
+            if (rowCells == null) {
+                throw new IndexOutOfBoundsException();
+            }
+            var row = new RowN(rowCells);
+            for (CollectExpression<Row, ?> expression : expressions) {
+                expression.setNextRow(row);
+            }
+            return args[0].value();
+        }
+    }
+
+    private final Supplier<OffsetDirection> FORWARD = () -> new OffsetDirection() {
```

Comment:
Might want to use regular inheritance here and have a `ForwardDirection` and `BackwardDirection` class.
The pattern here with anonymous classes addressed via a constant supplier definition is unusual.